armv7a9-zynq7000/serial.c

```diff
@@ -0,0 +1,343 @@
+		low_memcpy(buff + l, &serial->rxBuff[0], min(len - l, serial->rxHead));
+		cnt += min(len - l, serial->rxHead);
+	}
+	serial->rxTail = ((serial->rxTail + cnt) % BUFFER_SIZE);
+
+	low_sti();
+
+	return cnt;
+}
+
+
+int serial_write(unsigned int pn, const u8 *buff, u16 len)
+{
+	int l, cnt = 0;
+	serial_t *serial;
+
+	if (pn >= UARTS_MAX_CNT)
+		return ERR_ARG;
+
+	serial = &serial_common.serials[pn];
+
+	while (serial->txHead == serial->txTail && serial->tFull)
+		;
+
+	low_cli();
+	if (serial->txHead > serial->txTail)
+		l = min(serial->txHead - serial->txTail, len);
+	else
+		l = min(BUFFER_SIZE - serial->txTail, len);
+
+	low_memcpy(&serial->txBuff[serial->txTail], buff, l);
+	cnt = l;
+	if ((len > l) && (serial->txTail >= serial->txHead)) {
+		low_memcpy(serial->txBuff, buff + l, min(len - l, serial->txHead));
+		cnt += min(len - l, serial->txHead);
+	}
+
+	/* Initialize sending */
+	if (serial->txTail == serial->txHead)
+		*(serial->base + fifo) = serial->txBuff[serial->txHead];
+
+	serial->txTail = ((serial->txTail + cnt) % BUFFER_SIZE);
+
+	if (serial->txTail == serial->txHead)
+		serial->tFull = 1;
+
+	/* Enable TX FIFO empty irq */
+	 *(serial->base + ier) |= 1 << 3;
+	low_sti();
+
+	return cnt;
+}
+
+
+int serial_safewrite(unsigned int pn, const u8 *buff, u16 len)
+{
+	int l;
+
+	for (l = 0; len;) {
+		if ((l = serial_write(pn, buff, len)) < 0)
+			return ERR_MSG_IO;
+		buff += l;
+		len -= l;
+	}
+
+	return 0;
+}
+
+
+int serial_rxEmpty(unsigned int pn)
+{
+	serial_t *serial;
+
+	if (pn >= UARTS_MAX_CNT)
+		return ERR_ARG;
+
+	serial = &serial_common.serials[pn];
+
+	return serial->rxHead == serial->rxTail;
+}
+
+
+/* According to TRM:
+*  baud_rate = ref_clk / (bgen * (bdiv + 1))
+*  bgen: 2 - 65535
+*  bdiv: 4 - 255                             */
+static int serial_calcBaudarate(int pn, int baudrate)
+{
+	u32 bestDiff, diff;
+	u32 calcBaudrate;
+	u32 bdiv, bgen, bestBdiv = 4, bestBgen = 2;
+
+	serial_t *serial;
+
+	if (pn >= UARTS_MAX_CNT)
+		return ERR_ARG;
+
+	serial = &serial_common.serials[pn];
+
+	bestDiff = (u32)baudrate;
+
+	for (bdiv = 4; bdiv < 255; bdiv++) {
+		bgen = UART_REF_CLK / (baudrate * (bdiv + 1));
+
+		if (bgen < 2 || bgen > 65535)
+			continue;
+
+		calcBaudrate = UART_REF_CLK / (bgen * (bdiv + 1));
+
+		if (calcBaudrate > baudrate)
+			diff = calcBaudrate - baudrate;
+		else
+			diff = baudrate - calcBaudrate;
+
+		if (diff < bestDiff) {
+			bestDiff = diff;
+			bestBdiv = bdiv;
+			bestBgen = bgen;
+		}
+	}
+
+	*(serial->base + baudgen) = bestBgen;
+	*(serial->base + baud_rate_divider_reg0) = bestBdiv;
+
+	return ERR_NONE;
+}
+
+
+void serial_init(u32 baud, u32 *st)
+{
+	int i;
+	serial_t *serial;
+
+	static const struct {
+		volatile u32 *base;
+		u16 irq;
+	} info[] = {
+		{ UART0_BASE_ADDR, UART0_IRQ },
+		{ UART1_BASE_ADDR, UART1_IRQ }
+	};
+
+	/* TODO: initialiaze clocks and MIO pads
+	 *       zynq API has to be defined */
+	for (i = 0; i < UARTS_MAX_CNT; ++i) {
```

Comment:
Or you can ditch `UARTS_MAX_CNT` and use `sizeof(info) / sizeof(info[0])` instead.

Reply:
Ok but in that case some extra check before `for` loop should be added to make sure that `serial_common.serials` and `info` have the same size. Something like this:
```suggestion
      if ((sizeof(info) / sizeof(info[0])) != (sizeof(serial_common.serials) / sizeof(serial_common.serials[0]))
                return;
                
	for (i = 0; i < sizeof(info) / sizeof(info[0]); ++i) {
```

Reply:
Yeah, it would be a good idea... So let's stick to the `define ` then ;)

Reply:
I think checking at compile-time using `_Static_assert()` would be better.
Before `for` loop:
```
#define ARRAYSIZE(a)  (sizeof(a) / sizeof(a[0])

_Static_assert(ARRAYSIZE(info) == ARRAYSIZE(serial_common.serials), "size info != size serial_common.serials")
```
`#define ARRAYSIZE` is just to shorten the code.
Maybe a good solution would be introduce this array size definition and use it instead of `sizeof(a)  / sizeof(a[0])`.

Reply:
I think that checking arrays' size at the beginning is sufficient. Changes have been made.

Reply:
But why check at execution time if you can do it at compile time and if they are not equal there will be a compile error?
And so you get back from the function without any message.

Reply:
We avoid adding asserts.

Reply:
We avoid adding

Reply:
I don't understand why not use _Static_asset() in this situation, which I guess was created to solve just such cases among others.
The solution you propose is probably wrong.Why?
1. arrays have 2 elements each
```
if (2 != 2)
		return;
```
Condition is false, compiler skips it and there is a normal `serial_init()` function.

2. arrays have different number of elements e.g. info[] has 3 elements
```
if (3 != 2)
		return;
```
Condition is true, compiler optimizees this and `serial_init()` look like this
```
00000000 <serial_init>:
   0:   4770            bx      lr
```
`serial_init()` is empty and  does not work. You don't know why.

Now my suggestion.
Instaed of 
```
	if ((sizeof(info) / sizeof(info[0])) != (sizeof(serial_common.serials) / sizeof(serial_common.serials[0])))
		return;
```
you give
```
	_Static_assert((sizeof(info) / sizeof(info[0])) == (sizeof(serial_common.serials) / sizeof(serial_common.serials[0])), "serial_common.serial[] != info[]");
```
and at compilation you get the message:
```
CC  armv7a9-zynq7000/serial.c
armv7a9-zynq7000/serial.c: In function 'serial_init':
armv7a9-zynq7000/serial.c:288:2: error: static assertion failed: "serial_common.serial[] != info[]"
  288 |  _Static_assert((sizeof(info) / sizeof(info[0])) == (sizeof(serial_common.serials) / sizeof(serial_common.serials[0])), "serial_common.serial[] != info[]");
      |  ^~~~~~~~~~~~~~
```
And you know immediately what is wrong.

Reply:
I fully understand your approach. 
However, it is top-down decision to not use asserts so I follow the rules.
In this situation the best trade-off would be declare array info as `info[UARTS_MAX_CNT]` like you proposed in one of the comments.